frontend/pages/upload.vue

```diff
@@ -0,0 +1,407 @@
+  const isDuplicate = (fileState: FileState) => (fileNameAndSizeCounts.get(dupeKey(fileState)) ?? 0) > 1
+  return fileStates.value.map((fileState, index) => {
+    let icon = 'pi pi-spinner pi-spin'
+    if (fileState.status === FileStatus.Done) {
+      icon = 'pi pi-check-circle text-success'
+    } else if (fileState.status === FileStatus.Error) {
+      icon = 'pi pi-times text-error'
+    } else if (fileState.status === FileStatus.Selected) {
+      icon = 'pi pi-circle'
+    }
+    let otherError: string | undefined
+    // TODO(grady) validate this server side too.
+    if (fileState.file.name.length > 1000) {
+      otherError = 'Filename is too long (1000 characters max).'
+    } else if (fileState.file.size > 1028 * 1028 * 100) {
+      otherError = 'File is too large (100MB max).'
+    } else if (!fileState.file.name.endsWith('.csv')) {
+      otherError = 'File must be a csv file.'
+    } else if (isDuplicate(fileState)) {
+      otherError = 'This file may be a duplicate, consider removing it.'
+    }
+    return {
+      ...fileState,
+      index,
+      icon,
+      sizeStr: formatFileSize(fileState.file.size),
+      statusStr: tt(fileState.status),
+      key: `${fileState.file.name}-${index}`,
+      effectiveError: fileState.errorMessage ?? otherError,
+    }
+  })
+})
+const fileUploaderProps = computed(() => ({
+  disabled: isProcessing.value || allDone.value,
+  mode: 'basic',
+  auto: true,
+  multiple: true,
+  'custom-upload': true,
+  'choose-label': 'Add File(s)',
+}))
+const actionButtonLabel = computed(() => {
+  if (hasAnyState(FileStatus.Waiting)) {
+    return 'Waiting...'
+  }
+  if (hasAnyState(FileStatus.Uploading)) {
+    return 'Uploading...'
+  }
+  if (hasAnyState(FileStatus.Validating)) {
+    return 'Validating...'
+  }
+  if (hasAnyState(FileStatus.CleanUp)) {
+    return 'Cleaning Up...'
+  }
+  return 'Begin Upload'
+})
+const allDone = computed(() => hasAllState(FileStatus.Done) && fileStates.value.length > 0)
+
+const onSelect = (e: FileUploadUploaderEvent) => {
+  if (!e.files || e.files.length === 0) {
+    return
+  }
+  const fs = Array.isArray(e.files) ? e.files : [e.files]
+  fileStates.value.push(...fs.map((file) => ({
+    file,
+    fileName: file.name,
+    shortName: file.name.substring(file.name.lastIndexOf('/') + 1),
+    size: file.size,
+    status: FileStatus.Selected,
+  })))
+  setAllStates(FileStatus.Selected)
+}
+
+const startUpload = async () => {
+  errorCode.value = ''
+  errorMessage.value = ''
+  isProcessing.value = true
+  startedProcessing.value = true
+  setAllStates(FileStatus.Waiting)
+  const startPortfolioUploadResp = await pactaClient.startPortfolioUpload({
+    items: fileStates.value.map((fileState: FileState) => ({
+      file_name: fileState.file.name,
+      // TODO(grady) consider adding file size here as a validation step.
+      // TODO(grady) consider adding file type validation here.
+      // TODO(grady) consider adding duplicate file checking here.
```

Comment:
Isn't that happening above, or is that at a different point in the upload lifecycle?

Reply:
Good call - fixed that one!